maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java

```diff
@@ -408,34 +408,64 @@ private PluginVersionResult resolveFromProject( PluginVersionRequest request, Li
     }
 
     @SuppressWarnings( "unchecked" )
-    private ConcurrentMap<String, PluginVersionResult> getCache( SessionData data )
+    private ConcurrentMap<Key, PluginVersionResult> getCache( SessionData data )
     {
-        ConcurrentMap<String, PluginVersionResult> cache =
-                ( ConcurrentMap<String, PluginVersionResult> ) data.get( CACHE_KEY );
+        ConcurrentMap<Key, PluginVersionResult> cache =
+                ( ConcurrentMap<Key, PluginVersionResult> ) data.get( CACHE_KEY );
         while ( cache == null )
         {
             cache = new ConcurrentHashMap<>( 256 );
             if ( data.set( CACHE_KEY, null, cache ) )
             {
                 break;
             }
-            cache = ( ConcurrentMap<String, PluginVersionResult> ) data.get( CACHE_KEY );
+            cache = ( ConcurrentMap<Key, PluginVersionResult> ) data.get( CACHE_KEY );
         }
         return cache;
     }
 
-    private static String getKey( PluginVersionRequest request )
+    private static Key getKey( PluginVersionRequest request )
     {
-        StringBuilder sb = new StringBuilder();
-        sb.append( request.getGroupId() );
-        sb.append( ':' );
-        sb.append( request.getArtifactId() );
-        for ( RemoteRepository repository : request.getRepositories() )
+        return new Key( request.getGroupId(), request.getArtifactId(), request.getRepositories() );
+    }
+
+    static class Key
+    {
+        final String groupId;
+        final String artifactId;
+        final List<RemoteRepository> repositories;
+        final int hash;
+
+        Key( String groupId, String artifactId, List<RemoteRepository> repositories )
```

Comment:
The key makes the assumption/assertion that only one plugin version can be used at the same time?

Reply:
I don't think so, the key assumes a given request will always lead to the same result, which is the point of the cache.  There's no assumption about how the plugin version is used.